pkg/events/redis/store.go

```diff
@@ -0,0 +1,386 @@
+	*PubSub
+
+	historyTTL                time.Duration
+	entityHistoryCount        int
+	entityHistoryTTL          time.Duration
+	correlationIDHistoryCount int
+}
+
+func (ps *PubSubStore) eventDataKey(_ context.Context, uid string) string {
+	return ps.client.Key("event_data", uid)
+}
+
+// eventIndexKey returns the Redis for indexing events with the given correlationID.
+// Since correlation IDs are user input (and can therefore contain characters that
+// we may not want in Redis keys) we hash them.
+func (ps *PubSubStore) eventIndexKey(_ context.Context, correlationID string) string {
+	h := sha1.New()
+	h.Write([]byte(correlationID))
+	return ps.client.Key("event_index", base32.StdEncoding.EncodeToString(h.Sum(nil)))
+}
+
+func (ps *PubSubStore) eventStream(ctx context.Context, ids *ttnpb.EntityIdentifiers) string {
+	if ids == nil {
+		return ps.client.Key("event_stream")
+	}
+	return ps.client.Key("event_stream", ids.EntityType(), unique.ID(ctx, ids))
+}
+
+func (ps *PubSubStore) storeEvent(ctx context.Context, evt events.Event) error {
+	b, err := encodeEventData(evt)
+	if err != nil {
+		return err
+	}
+	_, err = ps.client.TxPipelined(ctx, func(tx redis.Pipeliner) error {
+		tx.Set(ps.ctx, ps.eventDataKey(evt.Context(), evt.UniqueID()), b, ps.historyTTL)
+		for _, cid := range evt.CorrelationIDs() {
+			key := ps.eventIndexKey(evt.Context(), cid)
+			tx.LPush(ps.ctx, key, evt.UniqueID())
+			tx.LTrim(ps.ctx, key, 0, int64(ps.correlationIDHistoryCount))
+			tx.Expire(ps.ctx, key, ps.historyTTL)
+		}
+		return nil
+	})
+	return ttnredis.ConvertError(err)
+}
+
+// loadEventData loads event data for every event in evts (by UniqueID)
+// and set additional fields in the events.
+func (ps *PubSubStore) loadEventData(ctx context.Context, evts ...*ttnpb.Event) error {
+	if len(evts) == 0 {
+		return nil
+	}
+	keys := make([]string, 0, len(evts))
+	for _, evt := range evts {
+		keys = append(keys, ps.eventDataKey(ctx, evt.UniqueID))
+	}
+	bs, err := ps.client.MGet(ctx, keys...).Result()
+	if err != nil {
+		return ttnredis.ConvertError(err)
+	}
+	for i, b := range bs {
+		switch b := b.(type) {
+		case nil:
+			continue // Event data deleted/expired.
+		case string:
+			if err = decodeEventData(b, evts[i]); err != nil {
+				log.FromContext(ctx).WithError(err).Warnf("Failed to decode event payload")
+				continue
+			}
+		default:
+			log.FromContext(ctx).Warnf("Invalid %T element in event payloads", b)
+			continue
+		}
+	}
+	return nil
+}
+
+// LoadEvent loads an event by its UID.
+func (ps *PubSubStore) LoadEvent(ctx context.Context, uid string) (*ttnpb.Event, error) {
+	data, err := ps.client.Get(ctx, ps.eventDataKey(ctx, uid)).Result()
+	if err != nil {
+		return nil, ttnredis.ConvertError(err)
+	}
+	var evtPB ttnpb.Event
+	if err = decodeEventData(data, &evtPB); err != nil {
+		return nil, err
+	}
+	return &evtPB, nil
+}
+
+func xMessageHasEventName(names ...string) func(msg redis.XMessage) bool {
+	if len(names) == 0 {
```

Comment:
Consider converting the list to a `map[string]struct{}` in order to avoid looping the whole thing during matches.